code_docs/Organization.md

```diff
@@ -0,0 +1,34 @@
+# Community firmware organization
+#### To work out
+- docs folder, (dev/bug_fixes) cleanup, consistency, how-to-use
+- what functionality is in hid vs gui and why?
+- why is colour in gui
+- why is waveform not in ui or views?
+
+### Naming conventions
+#### General observations
+- Naming conventions sometimes don't align between the public UI and the code. Song view being called session_view in code is a good example. These differences are also not documented. This makes it difficult for especially people new to the project to find their way
+- Some features could have more descriptive names. For example green/blue mode in Grid View are sometimes called launch and edit mode, but are in code and documentation referenced as green and blue. It would be helpful for both users and developers to refer to features by a descriptive name, both in code and documentation
+-  Some new menu-items - especially in `community firmware` - don't have descriptive names but are listed as "alternate [x] behavior", which puts the load on the user's memory to remember what they do. It would be helpful to have all menu names be descriptive
+-  At the same time, many menu items are longer than the display allows. This is probably not possible to avoid altogether, but navigating a menu with items that fit the display is significantly faster. It could pay off to rework menu items' naming to be both descriptive and short. For example: `stutter quantize` conveys as much information as `stutter rate quantize` while still fitting on the display. *(Note: some longer names from the community features menu will get shorter automatically when they are implemented in their dedicated spot in the regular menu, f.e. `grid view loop layer pads` will become `loop layer pads` when placed in the grid view menu)*
+
+### Code Cleanup
+#### Views
+- The current general strategy is to create global variables for all the main views and reuse them when activated. Is it safer/easier to refactor them to static properties?
+- None of the views inherit from view.h, which could be assumed would be the case from how the files are named. Instead all views inherit from `TimelineView`
+- `ClipNavigationTimelineView` is a `TimelineView` and only implements a few methods. `TimelineView` itself is only used as a base class for `ArrangerView`. This could probably be simplified into one class
+- **Proposal:**
+  - Combine `TimelineView` and `ClipNavigationTimelineView` to a new `View` class. This is a `RootUI` and will remain the base class of all other views. Shared behavior for views lives here
```

Comment:
In principal I agree with this refactoring approach of having a base view class from which all other view classes inherit.

To improve understanding of the implications of all this it would probably be worth searching for instances of getRootUI(), getCurrentUI(), changeRootUI(), &instrumentClipView, &automationView, etc. as it will give a better appreciation of how embedded those concepts are throughout the codebase.

In terms of refactoring view classes, it's definitely something we've talked about a lot. InstrumentClipView is the original big "catch all" class for interacting with melodic instrument and kit instrument clips. A lot of code could definitely be factored out to make the view code easier to comprehend.

If I give an example of automation view, which I initially built by taking a copy of instrument clip view and building the automation view functionality on top of, it's now a very huge beast of a class that needs to be refactored as it's just too big to manage now.

I want to refactor its current state so that automationView is a base class that other classes that take care of automationView specific functionality can inherit from / override.

e.g. Note Velocity Editor and Parameter Automation Editor are both part of the automationView class but they should be pulled out into their separate classes. 

When I started looking at it I started looking at splitting things out like this:

<img width="348" alt="Screenshot 2024-11-07 at 10 38 28 PM" src="https://github.com/user-attachments/assets/f52f401c-0ccd-41cf-a476-1f47b1698f1c">

Reply:
I definitely intend to spend some more time getting an understanding of the UI classes and how everything is implemented, before starting any refactoring work. Documenting the UI folder is next on the list. I feel we have a lot to gain here in both accessibility and possibly stability/speed.